invenio/modules/relationships/models.py

```diff
@@ -0,0 +1,163 @@
+## published by the Free Software Foundation; either version 2 of the
+## License, or (at your option) any later version.
+##
+## Invenio is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+## General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Invenio; if not, write to the Free Software Foundation, Inc.,
+## 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Database model for persisting relationships."""
+
+from uuid import uuid4
+
+from sqlalchemy_utils import UUIDType
+
+from invenio.ext.sqlalchemy import db, utils
+from invenio.modules.records.models import RecordMetadata
+
+from .errors import NoSuchRecordException, WrongModelNameException
+
+
+class Relationship(db.Model):
+
+    """Represent a graph edge."""
+
+    uuid = db.Column(UUIDType(binary=False), primary_key=True)
+    model_from = db.Column(db.String(255), index=True)
+    id_from = db.Column(db.Integer(unsigned=True), index=True)
+    link_type = db.Column(db.String(255), index=True)
+    model_to = db.Column(db.String(255), index=True)
+    id_to = db.Column(db.Integer(unsigned=True), index=True)
+
+    confidence = db.Column(db.Float, nullable=True)
+
+    def __init__(self, model_from, id_from, link_type, model_to, id_to):
+        """Initialize the relationship between two records.
+
+        Parameters
+        ----------
+        :param model_from: string
+            Model used by the source.
+        :param id_from: integer
+            Record id of the source.
+        :param link_type: string
+            Name of the relation.
+        :param model_to: string
+            Model used by the destination.
+        :param id_to: integer
+            Record id of the destination.
+        """
+        self.uuid = uuid4()
+        self.id_from = id_from
+        self.model_from = model_from
+        self.link_type = link_type
+        self.id_to = id_to
+        self.model_to = model_to
+
+    @db.validates('model_from')
+    def model_from_validation(self, key, model):
+        self._record_from = Node(self.id_from)
+        print self._record_from.get_model_names()
+        if model not in self._record_from.get_model_names():
+            raise WrongModelNameException("Wrong model for record %s" %
+                                          self.id_from)
+        return model
+
+    @db.validates('model_to')
+    def model_to_validation(self, key, model):
+        self._record_to = Node(self.id_to)
+        if model not in self._record_to.get_model_names():
+            raise WrongModelNameException("Wrong model for record %s" %
+                                          self.id_to)
+        return model
+
+    @utils.session_manager
+    def save(self):
+        db.session.add(self)
+
+
+class Node(object):
```

Comment:
Please check the original implementation. The idea behind was to allow linking between different models.